### ISSUE TITLE
Implementation of models

### ISSUE THREAD

**Issue by KongMingxi** [bowang-lab/scGPT#16]
Hi there, congrats and thanks for this excellent work! I am confused on the details of the implementation.

The paper shows that: 
"It consists of 12 stacked transformer blocks with 8 attention heads each."

But the code shows that:
`hyperparameter_defaults = dict(
    seed=42,
    dataset_name="PBMC_10K",
    do_train=True,
    load_model="save/scGPT_bc",
    mask_ratio=0.4,
    epochs=30,
    n_bins=51,
    GEPC=True,  # Masked value prediction for cell embedding
    ecs_thres=0.8,  # Elastic cell similarity objective, 0.0 to 1.0, 0.0 to disable
    dab_weight=1.0,
    lr=1e-4,
    batch_size=64,
    layer_size=128,
    nlayers=4,
    nhead=4,
    # if load model, batch_size, layer_size, nlayers, nhead will be ignored
    dropout=0.2,
    schedule_ratio=0.9,  # ratio of epochs for learning rate schedule
    save_eval_interval=5,
    log_interval=100,
    fast_transformer=True,
    pre_norm=False,
    amp=True,  # Automatic Mixed Precision
)`

I guess this means there are 4 blocks and 4 heads.
`    nlayers=4,
    nhead=4,`

I wonder if you can clarify this.

Best,

Mingxi

**Comment by subercui** [comment]
Yes, the pretrained model has 12 transformer blocks and 8 attention heads. As stated in the comments of the code snippet you copied, the `batch_size, layer_size, nlayers, nhead` will be updated if you load a pretrained model.  To be specific, these parameters will be later updated by the pretrained model args here

https://github.com/bowang-lab/scGPT/blob/6f1f5c9a8274b4305f58c77ed53446fe1c6d129d/examples/finetune_integration.py#L159

**Comment by subercui** [comment]
I also have a quick fix here to load the model args in `hyperparameter_defaults` when you are training from scratch without loading pretrained model

https://github.com/bowang-lab/scGPT/commit/dcaf1e382cd0b74e474783be94532eb915fff302#diff-76b1a198348cad2e01186fc1103195323e45fb75388dccbad08b00d9e16c0db8R163
